File: src/project-scanner.ts
import * as fs from 'fs';
import * as path from 'path';
import { RemoteConsole } from 'vscode-languageserver/node';

export interface ProjectData {
	components: Set<string>;
	componentProperties: Map<string, Set<string>>;
	componentFiles: Map<string, string>; // component -> file path
	fileComponents: Map<string, Set<string>>; // file path -> components
}

export class ProjectScanner {
	private projectData: ProjectData = {
		components: new Set(),
		componentProperties: new Map(),
		componentFiles: new Map(),
		fileComponents: new Map()
	};

	private watchers: Map<string, fs.FSWatcher> = new Map();

	constructor(
		private workspaceRoot: string,
		private console: RemoteConsole
	) {}

	async scanProject(): Promise<ProjectData> {
		this.console.log('[view.tree] Starting project scan...');
		
		// Reset project data
		this.projectData = {
			components: new Set(),
			componentProperties: new Map(),
			componentFiles: new Map(),
			fileComponents: new Map()
		};

		try {
			// Scan .view.tree files
			await this.scanViewTreeFiles();
			
			// Scan .ts files
			await this.scanTsFiles();

			// Setup file watchers
			this.setupFileWatchers();

			this.console.log(`[view.tree] Scan complete: ${this.projectData.components.size} components, ${this.projectData.componentProperties.size} components with properties`);
			this.console.log(`[view.tree] Components found: ${Array.from(this.projectData.components).join(', ')}`);

		} catch (error) {
			this.console.error(`[view.tree] Error during project scan: ${error}`);
		}

		return this.projectData;
	}

	private async scanViewTreeFiles(): Promise<void> {
		const viewTreeFiles = await this.findFiles('**/*.view.tree');
		this.console.log(`[view.tree] Found ${viewTreeFiles.length} .view.tree files`);

		for (const filePath of viewTreeFiles) {
			try {
				const content = await fs.promises.readFile(filePath, 'utf8');
				this.parseViewTreeFile(content, this.projectData, filePath);
			} catch (error) {
				this.console.log(`[view.tree] Error reading ${filePath}: ${error}`);
			}
		}
	}

	private async scanTsFiles(): Promise<void> {
		const tsFiles = await this.findFiles('**/*.ts');
		this.console.log(`[view.tree] Found ${tsFiles.length} .ts files`);

		// Limit to first 100 files for performance
		const limitedTsFiles = tsFiles.slice(0, 100);
		
		for (const file of limitedTsFiles) {
			try {
				const content = await fs.promises.readFile(file, 'utf8');
				this.parseTsFile(content, this.projectData, file);
			} catch (error) {
				this.console.log(`[view.tree] Error reading ${file}: ${error}`);
			}
		}
	}

	private async findFiles(pattern: string): Promise<string[]> {
		const files: string[] = [];
		
		const scanDirectory = async (dir: string): Promise<void> => {
			try {
				const entries = await fs.promises.readdir(dir, { withFileTypes: true });
				
				for (const entry of entries) {
					const fullPath = path.join(dir, entry.name);
					
					if (entry.isDirectory()) {
						// Skip node_modules and other common directories to ignore
						if (!entry.name.startsWith('.') && entry.name !== 'node_modules') {
							await scanDirectory(fullPath);
						}
					} else if (entry.isFile()) {
						if (pattern.includes('*.view.tree') && fullPath.endsWith('.view.tree')) {
							files.push(fullPath);
						} else if (pattern.includes('*.ts') && fullPath.endsWith('.ts') && !fullPath.endsWith('.d.ts')) {
							files.push(fullPath);
						}
					}
				}
			} catch (error) {
				// Ignore permission errors and continue
			}
		};

		await scanDirectory(this.workspaceRoot);
		return files;
	}

	private parseViewTreeFile(content: string, data: ProjectData, filePath: string): void {
		const lines = content.split('\n');
		let currentComponent: string | null = null;

		// Clear previous components for this file
		if (data.fileComponents.has(filePath)) {
			const previousComponents = data.fileComponents.get(filePath)!;
			for (const comp of previousComponents) {
				data.componentFiles.delete(comp);
			}
		}
		data.fileComponents.set(filePath, new Set());

		for (const line of lines) {
			const trimmed = line.trim();

			// Take only the first word from lines without indentation
			if (!line.startsWith('\t') && !line.startsWith(' ') && trimmed.startsWith('$')) {
				const firstWord = trimmed.split(/\s+/)[0];
				if (firstWord && firstWord.startsWith('$')) {
					currentComponent = firstWord;
					data.components.add(firstWord);
					data.componentFiles.set(firstWord, filePath);
					data.fileComponents.get(filePath)!.add(firstWord);
				
					if (!data.componentProperties.has(firstWord)) {
						data.componentProperties.set(firstWord, new Set());
					}
				}
			}

			// Look for properties (indented lines without <= and <=>)
			if (currentComponent) {
				const indentMatch = line.match(/^(\s+)([a-zA-Z_][a-zA-Z0-9_?*]*)\s*/);
				if (indentMatch && indentMatch[1] && indentMatch[1].length > 0 && !trimmed.includes('<=') && !trimmed.includes('<=>')) {
					const property = indentMatch[2];
					if (property && !property.startsWith('$') && property !== 'null' && property !== 'true' && property !== 'false') {
						data.componentProperties.get(currentComponent)!.add(property);
					}
				}

				// Look for properties in bindings: <= PropertyName
				const bindingMatch = trimmed.match(/<=\s+([a-zA-Z_][a-zA-Z0-9_?*]*)/);
				if (bindingMatch) {
					const property = bindingMatch[1];
					if (property && !property.startsWith('$')) {
						data.componentProperties.get(currentComponent)!.add(property);
					}
				}
			}
		}
	}

	private parseTsFile(content: string, data: ProjectData, filePath: string): void {
		// Look for all $ components in TypeScript files
		const componentMatches = content.match(/\$\w+/g);
		if (componentMatches) {
			// Clear previous components for this file
			if (data.fileComponents.has(filePath)) {
				const previousComponents = data.fileComponents.get(filePath)!;
				for (const comp of previousComponents) {
					if (data.componentFiles.get(comp) === filePath) {
						data.componentFiles.delete(comp);
					}
				}
			}
			data.fileComponents.set(filePath, new Set());

			for (const match of componentMatches) {
				data.components.add(match);
				// Only set file mapping if not already set by .view.tree file
				if (!data.componentFiles.has(match)) {
					data.componentFiles.set(match, filePath);
				}
				data.fileComponents.get(filePath)!.add(match);
			}
		}
	}

	private setupFileWatchers(): void {
		// Clean up existing watchers
		for (const watcher of this.watchers.values()) {
			watcher.close();
		}
		this.watchers.clear();

		try {
			// Watch for .view.tree files
			const viewTreeWatcher = fs.watch(this.workspaceRoot, { recursive: true }, (eventType, filename) => {
				if (filename && (filename.endsWith('.view.tree') || filename.endsWith('.ts'))) {
					const fullPath = path.join(this.workspaceRoot, filename);
					this.handleFileChange(eventType, fullPath);
				}
			});
			
			this.watchers.set('main', viewTreeWatcher);
		} catch (error) {
			this.console.error(`[view.tree] Error setting up file watchers: ${error}`);
		}
	}

	private async handleFileChange(eventType: string, filePath: string): Promise<void> {
		this.console.log(`[view.tree] File ${eventType}: ${filePath}`);
		
		try {
			if (eventType === 'change') {
				// File modified
				if (fs.existsSync(filePath)) {
					const content = await fs.promises.readFile(filePath, 'utf8');
					this.updateSingleFile(filePath, content);
				}
			} else {
				// File deleted or renamed
				this.removeSingleFile(filePath);
			}
		} catch (error) {
			this.console.error(`[view.tree] Error handling file change: ${error}`);
		}
	}

	public updateSingleFile(filePath: string, content: string): void {
		this.console.log(`[view.tree] Updating single file: ${filePath}`);
		
		try {
			if (filePath.endsWith('.view.tree')) {
				this.parseViewTreeFile(content, this.projectData, filePath);
			} else if (filePath.endsWith('.ts')) {
				this.parseTsFile(content, this.projectData, filePath);
			}
		} catch (error) {
			this.console.error(`[view.tree] Error updating file ${filePath}: ${error}`);
		}
	}

	private removeSingleFile(filePath: string): void {
		this.console.log(`[view.tree] File deleted: ${filePath}`);
		
		// Remove components that were defined in this file
		if (this.projectData.fileComponents.has(filePath)) {
			const components = this.projectData.fileComponents.get(filePath)!;
			for (const component of components) {
				if (this.projectData.componentFiles.get(component) === filePath) {
					this.projectData.components.delete(component);
					this.projectData.componentFiles.delete(component);
					this.projectData.componentProperties.delete(component);
				}
			}
			this.projectData.fileComponents.delete(filePath);
		}
	}

	public getProjectData(): ProjectData {
		return this.projectData;
	}

	public getComponentsStartingWith(prefix: string): string[] {
		return Array.from(this.projectData.components)
			.filter(component => component.startsWith(prefix))
			.sort();
	}

	public getPropertiesForComponent(component: string): string[] {
		const properties = this.projectData.componentProperties.get(component);
		return properties ? Array.from(properties).sort() : [];
	}

	public getAllProperties(): string[] {
		const allProperties = new Set<string>();
		for (const properties of this.projectData.componentProperties.values()) {
			for (const property of properties) {
				allProperties.add(property);
			}
		}
		return Array.from(allProperties).sort();
	}

	public getComponentFile(component: string): string | undefined {
		return this.projectData.componentFiles.get(component);
	}

	public dispose(): void {
		// Clean up file watchers
		for (const watcher of this.watchers.values()) {
			watcher.close();
		}
		this.watchers.clear();
	}
}

File: lsp/project-scanner.go
package main

import (
	"fmt"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"
)

type ProjectData struct {
	Components          map[string]bool            // Set of component names
	ComponentProperties map[string]map[string]bool // Map of component -> properties
	ComponentFiles      map[string]string          // Map of component -> file path
	FileComponents      map[string]map[string]bool // Map of file path -> components
	mutex               sync.RWMutex
}

func NewProjectData() *ProjectData {
	return &ProjectData{
		Components:          make(map[string]bool),
		ComponentProperties: make(map[string]map[string]bool),
		ComponentFiles:      make(map[string]string),
		FileComponents:      make(map[string]map[string]bool),
	}
}

type ProjectScanner struct {
	workspaceRoot string
	projectData   *ProjectData
}

func NewProjectScanner(workspaceRoot string) *ProjectScanner {
	return &ProjectScanner{
		workspaceRoot: workspaceRoot,
		projectData:   NewProjectData(),
	}
}

func (ps *ProjectScanner) ScanProject() error {
	log.Println("[view.tree] Starting project scan...")
	
	// Reset project data
	ps.projectData = NewProjectData()
	
	// Scan .view.tree files
	if err := ps.scanViewTreeFiles(); err != nil {
		log.Printf("[view.tree] Error scanning view.tree files: %v", err)
	}
	
	// Scan .ts files
	if err := ps.scanTsFiles(); err != nil {
		log.Printf("[view.tree] Error scanning ts files: %v", err)
	}
	
	ps.projectData.mutex.RLock()
	componentCount := len(ps.projectData.Components)
	propertiesCount := len(ps.projectData.ComponentProperties)
	ps.projectData.mutex.RUnlock()
	
	log.Printf("[view.tree] Scan complete: %d components, %d components with properties", componentCount, propertiesCount)
	
	var componentNames []string
	ps.projectData.mutex.RLock()
	for component := range ps.projectData.Components {
		componentNames = append(componentNames, component)
	}
	ps.projectData.mutex.RUnlock()
	
	if len(componentNames) > 0 {
		sort.Strings(componentNames)
		log.Printf("[view.tree] Components found: %s", strings.Join(componentNames, ", "))
	}
	
	return nil
}

func (ps *ProjectScanner) scanViewTreeFiles() error {
	viewTreeFiles, err := ps.findFiles("**/*.view.tree")
	if err != nil {
		return fmt.Errorf("failed to find view.tree files: %w", err)
	}
	
	log.Printf("[view.tree] Found %d .view.tree files", len(viewTreeFiles))
	
	for _, filePath := range viewTreeFiles {
		content, err := os.ReadFile(filePath)
		if err != nil {
			log.Printf("[view.tree] Error reading %s: %v", filePath, err)
			continue
		}
		
		ps.parseViewTreeFile(string(content), filePath)
	}
	
	return nil
}

func (ps *ProjectScanner) scanTsFiles() error {
	tsFiles, err := ps.findFiles("**/*.ts")
	if err != nil {
		return fmt.Errorf("failed to find ts files: %w", err)
	}
	
	log.Printf("[view.tree] Found %d .ts files", len(tsFiles))
	
	// Limit to first 100 files for performance
	if len(tsFiles) > 100 {
		tsFiles = tsFiles[:100]
	}
	
	for _, filePath := range tsFiles {
		content, err := os.ReadFile(filePath)
		if err != nil {
			log.Printf("[view.tree] Error reading %s: %v", filePath, err)
			continue
		}
		
		ps.parseTsFile(string(content), filePath)
	}
	
	return nil
}

func (ps *ProjectScanner) findFiles(pattern string) ([]string, error) {
	var files []string
	
	err := filepath.WalkDir(ps.workspaceRoot, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // Skip errors and continue
		}
		
		if d.IsDir() {
			// Skip hidden directories and node_modules
			if strings.HasPrefix(d.Name(), ".") || d.Name() == "node_modules" {
				return filepath.SkipDir
			}
			return nil
		}
		
		if strings.Contains(pattern, "*.view.tree") && strings.HasSuffix(path, ".view.tree") {
			files = append(files, path)
		} else if strings.Contains(pattern, "*.ts") && strings.HasSuffix(path, ".ts") && !strings.HasSuffix(path, ".d.ts") {
			files = append(files, path)
		}
		
		return nil
	})
	
	return files, err
}

func (ps *ProjectScanner) parseViewTreeFile(content, filePath string) {
	lines := strings.Split(content, "\n")
	var currentComponent string
	
	ps.projectData.mutex.Lock()
	defer ps.projectData.mutex.Unlock()
	
	// Clear previous components for this file
	if components, exists := ps.projectData.FileComponents[filePath]; exists {
		for comp := range components {
			if ps.projectData.ComponentFiles[comp] == filePath {
				delete(ps.projectData.ComponentFiles, comp)
			}
		}
	}
	ps.projectData.FileComponents[filePath] = make(map[string]bool)
	
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		
		// Take only the first word from lines without indentation
		if !strings.HasPrefix(line, "\t") && !strings.HasPrefix(line, " ") && strings.HasPrefix(trimmed, "$") {
			fields := strings.Fields(trimmed)
			if len(fields) > 0 {
				firstWord := fields[0]
				if strings.HasPrefix(firstWord, "$") {
					currentComponent = firstWord
					ps.projectData.Components[firstWord] = true
					ps.projectData.ComponentFiles[firstWord] = filePath
					ps.projectData.FileComponents[filePath][firstWord] = true
					
					if _, exists := ps.projectData.ComponentProperties[firstWord]; !exists {
						ps.projectData.ComponentProperties[firstWord] = make(map[string]bool)
					}
				}
			}
		}
		
		// Look for properties (indented lines without <= and <=>)
		if currentComponent != "" {
			indentMatch := regexp.MustCompile(`^(\s+)([a-zA-Z_][a-zA-Z0-9_?*]*)\s*`).FindStringSubmatch(line)
			if len(indentMatch) > 2 && len(indentMatch[1]) > 0 && 
			   !strings.Contains(trimmed, "<=") && !strings.Contains(trimmed, "<=>") {
				property := indentMatch[2]
				if property != "" && !strings.HasPrefix(property, "$") && 
				   property != "null" && property != "true" && property != "false" {
					ps.projectData.ComponentProperties[currentComponent][property] = true
				}
			}
			
			// Look for properties in bindings: <= PropertyName
			bindingMatch := regexp.MustCompile(`<=\s+([a-zA-Z_][a-zA-Z0-9_?*]*)`).FindStringSubmatch(trimmed)
			if len(bindingMatch) > 1 {
				property := bindingMatch[1]
				if property != "" && !strings.HasPrefix(property, "$") {
					ps.projectData.ComponentProperties[currentComponent][property] = true
				}
			}
		}
	}
}

func (ps *ProjectScanner) parseTsFile(content, filePath string) {
	// Look for all $ components in TypeScript files
	componentRegex := regexp.MustCompile(`\$\w+`)
	matches := componentRegex.FindAllString(content, -1)
	
	if len(matches) == 0 {
		return
	}
	
	ps.projectData.mutex.Lock()
	defer ps.projectData.mutex.Unlock()
	
	// Clear previous components for this file
	if components, exists := ps.projectData.FileComponents[filePath]; exists {
		for comp := range components {
			if ps.projectData.ComponentFiles[comp] == filePath {
				delete(ps.projectData.ComponentFiles, comp)
			}
		}
	}
	ps.projectData.FileComponents[filePath] = make(map[string]bool)
	
	for _, match := range matches {
		ps.projectData.Components[match] = true
		// Only set file mapping if not already set by .view.tree file
		if _, exists := ps.projectData.ComponentFiles[match]; !exists {
			ps.projectData.ComponentFiles[match] = filePath
		}
		ps.projectData.FileComponents[filePath][match] = true
	}
}

func (ps *ProjectScanner) UpdateSingleFile(filePath, content string) {
	log.Printf("[view.tree] Updating single file: %s", filePath)
	
	if strings.HasSuffix(filePath, ".view.tree") {
		ps.parseViewTreeFile(content, filePath)
	} else if strings.HasSuffix(filePath, ".ts") {
		ps.parseTsFile(content, filePath)
	}
}

func (ps *ProjectScanner) GetProjectData() *ProjectData {
	return ps.projectData
}

func (ps *ProjectScanner) GetComponentsStartingWith(prefix string) []string {
	ps.projectData.mutex.RLock()
	defer ps.projectData.mutex.RUnlock()
	
	var components []string
	for component := range ps.projectData.Components {
		if strings.HasPrefix(component, prefix) {
			components = append(components, component)
		}
	}
	
	sort.Strings(components)
	return components
}

func (ps *ProjectScanner) GetPropertiesForComponent(component string) []string {
	ps.projectData.mutex.RLock()
	defer ps.projectData.mutex.RUnlock()
	
	properties, exists := ps.projectData.ComponentProperties[component]
	if !exists {
		return []string{}
	}
	
	var result []string
	for property := range properties {
		result = append(result, property)
	}
	
	sort.Strings(result)
	return result
}

func (ps *ProjectScanner) GetAllProperties() []string {
	ps.projectData.mutex.RLock()
	defer ps.projectData.mutex.RUnlock()
	
	allProperties := make(map[string]bool)
	for _, properties := range ps.projectData.ComponentProperties {
		for property := range properties {
			allProperties[property] = true
		}
	}
	
	var result []string
	for property := range allProperties {
		result = append(result, property)
	}
	
	sort.Strings(result)
	return result
}

func (ps *ProjectScanner) GetComponentFile(component string) string {
	ps.projectData.mutex.RLock()
	defer ps.projectData.mutex.RUnlock()
	
	return ps.projectData.ComponentFiles[component]
}

// GetComponents returns all components
func (ps *ProjectScanner) GetComponents() []string {
	ps.projectData.mutex.RLock()
	defer ps.projectData.mutex.RUnlock()
	
	var components []string
	for component := range ps.projectData.Components {
		components = append(components, component)
	}
	
	sort.Strings(components)
	return components
}

// HasComponent checks if a component exists
func (ps *ProjectScanner) HasComponent(component string) bool {
	ps.projectData.mutex.RLock()
	defer ps.projectData.mutex.RUnlock()
	
	return ps.projectData.Components[component]
}